### PR TITLE
Add styling and additional logic to search for users

### LINKED ISSUE #98: [User Story] Search For Contacts In Mobile Version
# Description
As a user I want to search for contacts on my mobile, too.

# Dependencies
- [x] #97 

# Conditions of satisfaction
- [ ] the design follows the wireframe below
- [ ] there is a field to search for contacts
- [ ] only related contacts (= search results) are shown
- [ ] search results are alphabetically sorted

Mobile Contacts v2

### LINKED ISSUE #127: [Process Part] Search For Users
# Description
Als Nutzer möchte ich andere Benutzer über eine Suche finden können.

# Dependencies
- [ ] #126

# Conditions of satisfaction
- [ ] im Header gibt es ein Suchsymbol, auf das ich klicken kann und dann werden mir nur noch Benutzerkarten angezeigt, die zu der aktuellen Suche passen

### PR DESCRIPTION
**Pull request by Alexander-Dubrawski**
This PR closes #127 and closes #98 .

In the navbar is now an icon that is redirecting the user to the search page. 

The styling is now looking as follows: 

Desktop:
 
Screenshot 2021-01-29 at 22 31 56

Mobile: 

Screenshot 2021-01-29 at 22 08 54

### CODE REVIEW

File: app/controllers/users_controller.rb
Comment (Paula-Kli) [inline app/controllers/users_controller.rb]: why are you not using the method from the Users model anymore? I don't think this is a good way to implement searching since you are using a method from the users helper but nevertheless have to tell which record you want to search in.
`User.search(params[:search]).where.not(id: current_user.id)` this line gave you all users which match the search parameter and which are not the current user

File: app/controllers/users_controller.rb
Comment (Paula-Kli) [inline app/controllers/users_controller.rb]: if you are just turning that to `true` to give that to the method, why are you not putting `true` into the method call?

File: app/controllers/users_controller.rb
Comment (Paula-Kli) [inline app/controllers/users_controller.rb]: I think that variable could be named `@current_user` to be more precise

File: app/views/layouts/_flash_messages.html.erb
Comment (Paula-Kli) [inline app/views/layouts/_flash_messages.html.erb]: what does this change mean? why is this changed in the "change search" pr?

File: app/views/users/search.html.erb
Comment (Paula-Kli) [inline app/views/users/search.html.erb]: is there a reason, why the logic is changed? should I first see my contacts and later other people?

I would not call one contact just user but `contact` to know why they are handled differently

when #158  ist merged you can just call `<%= render partial: 'users/card', locals: { user: contact, mode: :overview } %>` here to get the expected card with the social accounts

File: app/views/users/search.html.erb
Comment (Paula-Kli) [inline app/views/users/search.html.erb]: when #158  ist merged you can simply use `<%= render partial: 'users/card', locals: { user: user, mode: :add } %>` to get the expected card with a `+` button and a tooltip

File: config/routes.rb
Comment (Paula-Kli) [inline config/routes.rb]: why is this line added when the same is done in line 14?

File: app/controllers/users_controller.rb
Comment (Paula-Kli) [inline app/controllers/users_controller.rb]: please use `current_user.sent_contact_request?` here because we not only want to send no second request to a user that is already our contact but also not to someone with a pending request

**Comment by Paula-Kli** [comment]
should the empty search page maybe display all available users?

File: app/controllers/users_controller.rb
Comment (felixauringer) [reply app/controllers/users_controller.rb]: Because it is currently not necessary to switch wildcards off, I would completely omit this variable and always use wildcards.

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: I would prefer changing code not the config of rubocop. Why should we use it at all if we are turning it off when it detects opportunities for improvements?

File: app/assets/stylesheets/users.scss
Comment (felixauringer) [inline app/assets/stylesheets/users.scss]: missing newline

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: Why do you change the authorization settings for this controller? I couldn't find it specified in the linked ticket and I think @LinqLover did really good work when introducing proper authentication.

File: app/controllers/users_controller.rb
Comment (felixauringer) [reply app/controllers/users_controller.rb]: I would strongly support this. The search logic built by Scrumbledore's Army is reviewed and well-tested. I do not see any reason to rebuild this logic. And I would also prefer the approach of the existing solution over your new approach.

File: app/helpers/users_helper.rb
Comment (felixauringer) [inline app/helpers/users_helper.rb]: You can't just disable Rubocop wheresoever it complains about something :sweat:

File: app/views/layouts/_navbar.html.erb
Comment (felixauringer) [inline app/views/layouts/_navbar.html.erb]: please use path helpers

File: app/views/users/search.html.erb
Comment (felixauringer) [reply app/views/users/search.html.erb]: That is now merged :+1:

File: app/views/users/search.html.erb
Comment (felixauringer) [inline app/views/users/search.html.erb]: This is almost the same code as for contacts above

**Review by felixauringer — CHANGES_REQUESTED** [review]
Remove the `.DS_Store` file please. You could also add it to your global gitignore on your computer (not in the project's gitignore) to prevent committing this file in all git projects you are working on.

You should merge `dev` into this branch because it already contains a lot of changes that you tried to implement here. I would always wait for PRs to be merged instead of implementing them yourself. That duplicates work and increases the size of PRs so reviewing gets a lot harder.

So I think all the things mentioned in `Additional implementation` in the PR description should not be in this branch. For example the logout button: If you notice that it is missing, you could open a ticket, fix it with maybe 5 lines of code, write one test for it and the PR could be merged in one day. But if you include something like this in a PR as big as this one, it takes forever until users can log out again.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: Hm... You are right. I don't know what happened there 🤔 . I will use the search logic from before.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: I replaced the search logic with the one from Scrumbledore's Army. So we are not using this variable anymore.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: Same as above.

File: app/views/layouts/_flash_messages.html.erb
Comment (Alexander-Dubrawski) [reply app/views/layouts/_flash_messages.html.erb]: It was just a quick fix to be able to continue implementing. But after the merge with the master, I replaced the code with the new code from there.

File: app/views/users/search.html.erb
Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: > is there a reason, why the logic is changed? should I first see my contacts and later other people?

Yes, that is the desired behavior.

Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: I would like to use the new design, especially since it is a pretty nice generic solution. But after some testing (Iphone 6/7/8), I am a little bit conflicted. The links don't seem to be placed correctly:  

Screenshot 2021-02-03 at 20 59 19

I just changed the code to:

```rails
      <% @users.each do |user| %>
             <%= render partial: 'users/card', locals: { user: user, mode: :overview } %>
     <% end %>
```

Am I doing something wrong? 😞

File: app/views/users/search.html.erb
Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: Please have a look at the conversation above.

File: config/routes.rb
Comment (Alexander-Dubrawski) [reply config/routes.rb]: We need this route if we try to reload the search page. Not sure why it is not working with the one defined in line 14 though, to be honest.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: Removed this search logic.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: In my opinion, sometimes methods will be more complex than 10 lines of code. But I am programming quite a lot of c++ lately so maybe I just have a different subjective view on this matter. Since I replaced this method with the one of  Scrumbledore's Army I don't need to disable rubocop here.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: Not sure why I did it. I removed it.

File: app/views/users/search.html.erb
Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: I know 🙈 . Just wanted to wait for the implementation of  #158 . Otherwise, I would have implemented some genric partials.

File: app/views/users/search.html.erb
Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: If you can't find a quick fix I could try to adjust it with some CSS.

Comment (Paula-Kli) [reply app/views/users/search.html.erb]: I dont think you are doing somethin wrong and for me with firefox and iPhone 6/7/8 it looks like this:
grafik

Comment (Paula-Kli) [reply app/views/users/search.html.erb]: I changed the code, to the desired partial but I am on your branch

File: config/routes.rb
Comment (Paula-Kli) [reply config/routes.rb]: did you try using the path helper `search_users_path` as mentioned below it is not just `/search` but `users/search` and due to that the correct path is `search_users_path`

File: app/views/layouts/_flash_messages.html.erb
Comment (rheaSPK) [inline app/views/layouts/_flash_messages.html.erb]: Please add new line

File: app/views/layouts/_footer.html.erb
Comment (rheaSPK) [inline app/views/layouts/_footer.html.erb]: please put this a seperate css file like `mobile.scss`

**Review by rheaSPK — COMMENTED** [review]
Please add the new card style, it's already merged and please fix CI-Errors, somethings not right with some of your code...

File: app/views/layouts/_footer.html.erb
Comment (felixauringer) [inline app/views/layouts/_footer.html.erb]: missing newline

File: app/views/layouts/_navbar.html.erb
Comment (felixauringer) [inline app/views/layouts/_navbar.html.erb]: missing newline

File: app/views/users/search.html.erb
Comment (felixauringer) [reply app/views/users/search.html.erb]: #158 has been merged now

File: spec/features/user_search_spec.rb
Comment (felixauringer) [inline spec/features/user_search_spec.rb]: Why did you add this test? Exactly this behaviour is tested in the line 55 (before your changes) / 62 (after your changes).
If you want to check that the first name is visible, then I would add this expectation to all tests below

File: spec/views/users/search.html.erb_spec.rb
Comment (felixauringer) [inline spec/views/users/search.html.erb_spec.rb]: What's your thought behind the changes in this test? You have renamed the variable in the test and added the existing test again with the renamed variable but that shouldn't change the behaviour of the test. I would therefore restore this file to its current state on dev

File: app/assets/stylesheets/users.scss
Comment (felixauringer) [inline app/assets/stylesheets/users.scss:4]: Is this file still necessary with #158 merged?

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: why?

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: What does this do? I would expect the symbol of a method name here that would be executed before the specified actions

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: ?
```suggestion
```

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: This change should not be necessary if you would leave `before_action :authorize` this would not be necessary

File: app/controllers/users_controller.rb
Comment (felixauringer) [inline app/controllers/users_controller.rb]: This change should not be necessary if you would leave before_action :authorize this would not be necessary

File: app/views/users/search.html.erb
Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: @Paula-Kli there are a few problems with the cards that catch my eye:

* Why are we using tables to style the cards? Tables were used in the early days of CSS. Now usually flex or grids are used for that. With flex, you are able to place the content much nicer and evenly. For example, you can distribute the handles evenly. I think the problem that we saw is caused by using tables.
* Why are we using cards from Bootstrap. Bootstrap cards usually have a different layout. If we want to simply display a shadow and have round corners we can  use basic CSS

I could quickly adjust the code of your partials so that the problems are solved if it is ok for you 🙃 . We already have all the CSS.

File: app/assets/stylesheets/users.scss
Comment (Alexander-Dubrawski) [reply app/assets/stylesheets/users.scss:4]: Yes, it is. We are using some styles for the search form.

File: config/locales/en.yml
Comment (frcroth) [inline config/locales/en.yml]: Use placeholder instead of place_holder

File: app/views/layouts/_footer.html.erb
Comment (frcroth) [inline app/views/layouts/_footer.html.erb]: This comment does not seem valid

File: app/views/layouts/_footer.html.erb
Comment (frcroth) [inline app/views/layouts/_footer.html.erb]: Remove

File: app/views/layouts/_footer.html.erb
Comment (frcroth) [inline app/views/layouts/_footer.html.erb]: Replace root_path with chat_path

File: app/views/layouts/_footer.html.erb
Comment (frcroth) [inline app/views/layouts/_footer.html.erb]: What is the reason for the changes in this file? It does not seem to be related to the topic of the PR. You could remove this here/ move it to another PR to reduce complexity

**Review by frcroth — CHANGES_REQUESTED** [review]

File: app/views/layouts/_footer.html.erb
Comment (Alexander-Dubrawski) [reply app/views/layouts/_footer.html.erb]: Maybe the following screenshot describes it the best: 

Screenshot 2021-02-05 at 17 09 16

The link of the setting page is covering the whole page.

Comment (Alexander-Dubrawski) [reply app/views/layouts/_footer.html.erb]: To solve this problem I simply wrap the link inside a label.

Comment (frcroth) [reply app/views/layouts/_footer.html.erb]: Sorry I don't have access to a PC right now. Can you reproduce this @felixauringer?

File: config/routes.rb
Comment (Alexander-Dubrawski) [reply config/routes.rb]: Now it works, thanks 😄 .

File: spec/views/users/search.html.erb_spec.rb
Comment (Alexander-Dubrawski) [reply spec/views/users/search.html.erb_spec.rb]: We have *users* and *users_to_add* and we want to check that both groups are being rendered. That's why we have two test cases.

File: spec/features/user_search_spec.rb
Comment (Alexander-Dubrawski) [reply spec/features/user_search_spec.rb]: I removed the test.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: I added `before_action :authorize` again.

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: I changed it back

File: app/controllers/users_controller.rb
Comment (Alexander-Dubrawski) [reply app/controllers/users_controller.rb]: changed it back

File: app/views/layouts/_footer.html.erb
Comment (Alexander-Dubrawski) [reply app/views/layouts/_footer.html.erb]: @frcroth is it okay if I leave it in this PR? I know it's not the best solution, but at the moment I prefer the fastest one.

File: spec/features/user_search_spec.rb
Comment (felixauringer) [inline spec/features/user_search_spec.rb]: Is this user still needed? It is the only change in this file, so I assume that it can be omitted?
```suggestion
```

File: spec/views/users/search.html.erb_spec.rb
Comment (felixauringer) [reply spec/views/users/search.html.erb_spec.rb]: That's a good idea, but that is not what the test does :sweat_smile:

Comment (felixauringer) [reply spec/views/users/search.html.erb_spec.rb]: `assign(:users, contacts)` is like executing `@users = contacts` in the controller before rendering the view (    assign(:users, contacts))

File: spec/views/users/search.html.erb_spec.rb
Comment (felixauringer) [inline spec/views/users/search.html.erb_spec.rb]:
```suggestion
  let(:user) { FactoryBot.create :user }
```

File: spec/views/users/search.html.erb_spec.rb
Comment (felixauringer) [inline spec/views/users/search.html.erb_spec.rb]:
```suggestion
    assign(:users, contacts + users)
```

File: spec/views/users/search.html.erb_spec.rb
Comment (felixauringer) [inline spec/views/users/search.html.erb_spec.rb]:
```suggestion
    sign_in user
```

**Review by felixauringer — COMMENTED** [review]
I added some suggestions that assign the instance variables in the view as the users controller would do

File: app/views/layouts/_footer.html.erb
Comment (felixauringer) [reply app/views/layouts/_footer.html.erb]: image

I'm on your branch @Alexander-Dubrawski but I resetted `app/views/layouts/_footer.html.erb` to its state on the `dev` branch and everything seems to be fine without your changes in this file. I'm not quite sure what the problem is :sweat_smile:

Comment (Alexander-Dubrawski) [reply app/views/layouts/_footer.html.erb]: That's confusing. I have the same problem on dev.  But I will double-check.

Comment (Alexander-Dubrawski) [reply app/views/layouts/_footer.html.erb]: Well..., somehow it got magically fixed. I can't reproduce it anymore either. Maybe one of the last merges fixed it. Its getting hard to keep track of.

File: spec/views/users/search.html.erb_spec.rb
Comment (Alexander-Dubrawski) [reply spec/views/users/search.html.erb_spec.rb]: I just changed the code back to the dev status.

File: app/views/users/search.html.erb
Comment (Alexander-Dubrawski) [reply app/views/users/search.html.erb]: will be solved with #189 .

**Comment by felixauringer** [comment]
The code looks great now! The PR has also a size that is very good reviewable. Just one question:
```ruby
  def search
    @users = User.search(params[:search]).where.not(id: current_user.id)
    @users_to_add = @users.reject do |user|
      current_user.sent_contact_request?(user)
    end
  end
```
It looks like `users_to_add` is a subset of `users`. Did you try out whether users that aren't contacts of (or requested by) the current user are displayed twice?
If yes, I would suggest something like the following
```ruby
  def search
    other_users = User.search(params[:search]).where.not(id: current_user.id)
    @users_to_add = other_users.reject do |user|
      current_user.sent_contact_request?(user)
    end
    @users = other_users.select do |user|
      current_user.sent_contact_request?(user)
    end
  end
```

**Comment by Alexander-Dubrawski** [comment]
> The code looks great now! The PR has also a size that is very good reviewable. 

Sorry for this chaotic PR of mine 🙈 . I really don't know why I implemented stuff that was already existing. But thanks for all your feedback and patience. 
  

> It looks like `users_to_add` is a subset of `users`. Did you try out whether users that aren't contacts of (or requested by) the current user are displayed twice?

You are absolutely right. Good eye. I adjusted the code.

**Review by felixauringer — APPROVED** [review]
Nice to finally finish this :tada: I'm sorry for having so much comments but thank you for implementing all of it :)

**Comment by frcroth** [comment]
image
This is what /users/search looks for me on the deployed version

**Comment by frcroth** [comment]
@Alexander-Dubrawski

**Comment by Alexander-Dubrawski** [comment]
> This is what /users/search looks for me on the deployed version

I can not reproduce the error. @frcroth what exactly did you do? (For example: login, click the search icon, ...)

**Comment by felixauringer** [comment]
> image
> This is what /users/search looks for me on the deployed version

Same for me, I will try to investigate the reason

**Comment by Alexander-Dubrawski** [comment]
> Same for me, I will try to investigate the reason

@felixauringer thanks that would be really nice. If you can not find the error, please ping me and I can also have a look this evening.

**Comment by felixauringer** [comment]
Works locally for me, so I will have a look at the heroku logs

**Comment by felixauringer** [comment]
The relevant part is

```text
ActionView::Template::Error (Can't resolve image into URL: to_model delegated to attachment, but attachment is nil):

2021-02-06T15:00:46.081546+00:00 app[web.1]:      3:     <div class="container">

2021-02-06T15:00:46.081546+00:00 app[web.1]:      4:       <div class="row">

2021-02-06T15:00:46.081547+00:00 app[web.1]:      5:         <div class="col-xs">

2021-02-06T15:00:46.081548+00:00 app[web.1]:      6:           <%= image_tag user.avatar, height: 65, width: 65, class: "rounded-circle" %>

2021-02-06T15:00:46.081548+00:00 app[web.1]:      7:         </div>

2021-02-06T15:00:46.081554+00:00 app[web.1]:      8:         <a class="col-8" href=<%= user_path(user) %>>

2021-02-06T15:00:46.081555+00:00 app[web.1]:      9:           <h5 class="card-title"><%= user.display_name %></h5>

2021-02-06T15:00:46.081556+00:00 app[web.1]:   

2021-02-06T15:00:46.081556+00:00 app[web.1]:  app/views/users/_card.html.erb:6

2021-02-06T15:00:46.081556+00:00 app[web.1]:  app/views/users/search.html.erb:14

2021-02-06T15:00:46.081557+00:00 app[web.1]:  app/views/users/search.html.erb:13:in `each'

2021-02-06T15:00:46.081557+00:00 app[web.1]:  app/views/users/search.html.erb:13
```

**Comment by felixauringer** [comment]
So again problems with the avatars. This PR did not touch `_card.html.erb` so this problem was likely there before merging this one

**Comment by felixauringer** [comment]
That's executed on production data:

```ruby
User.all.select { |user| user.avatar.nil? }.length
D, [2021-02-06T15:08:13.259421 #4] DEBUG -- :   User Load (1.8ms)  SELECT "users".* FROM "users"
=> 0
```
So the migration worked as expected

**Comment by felixauringer** [comment]
I added an issue for this: #191

**Pull request merged**
